completion/tcolorbox.cwl

```diff
@@ -321,81 +357,89 @@ all
+\renewtcboxfit{cmd}[args][default]{options%keyvals}#*
+\renewtcboxfit[init options]{cmd}{options%keyvals}#*
+\renewtcboxfit[init options]{cmd}[args]{options%keyvals}#*
+\renewtcboxfit[init options]{cmd}[args][default]{options%keyvals}#*
 \tcbfontsize{factor}
 \tcbfitdim
 \tcbfitsteps
 
 # << xparse >>
 \DeclareTColorBox{envname}{specification}{options%keyvals}#N
-\DeclareTColorBox[init options%keyvals]{envname}{specification}{options%keyvals}#*N
+\DeclareTColorBox[init options]{envname}{specification}{options%keyvals}#*N
 \NewTColorBox{envname}{specification}{options%keyvals}#N
-\NewTColorBox[init options%keyvals]{envname}{specification}{options%keyvals}#*N
+\NewTColorBox[init options]{envname}{specification}{options%keyvals}#*N
 \RenewTColorBox{envname}{specification}{options%keyvals}
-\RenewTColorBox[init options%keyvals]{envname}{specification}{options%keyvals}#*
+\RenewTColorBox[init options]{envname}{specification}{options%keyvals}#*
 \ProvideTColorBox{envname}{specification}{options%keyvals}#N
-\ProvideTColorBox[init options%keyvals]{envname}{specification}{options%keyvals}#*N
-\DeclareTotalTColorBox{\name}{specification}{options%keyvals}{content}
-\DeclareTotalTColorBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
-\NewTotalTColorBox{\name}{specification}{options%keyvals}{content}
-\NewTotalTColorBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
-\RenewTotalTColorBox{\name}{specification}{options%keyvals}{content}
-\RenewTotalTColorBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
-\ProvideTotalTColorBox{\name}{specification}{options%keyvals}{content}
-\ProvideTotalTColorBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
-\DeclareTCBox{\name}{specification}{options%keyvals}
-\DeclareTCBox[init options%keyvals]{\name}{specification}{options%keyvals}#*
-\NewTCBox{\name}{specification}{options%keyvals}
-\NewTCBox[init options%keyvals]{\name}{specification}{options%keyvals}#*
-\RenewTCBox{\name}{specification}{options%keyvals}
-\RenewTCBox[init options%keyvals]{\name}{specification}{options%keyvals}#*
-\ProvideTCBox{\name}{specification}{options%keyvals}
-\ProvideTCBox[init options%keyvals]{\name}{specification}{options%keyvals}#*
-\DeclareTotalTCBox{\name}{specification}{options%keyvals}{content}
-\DeclareTotalTCBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
-\NewTotalTCBox{\name}{specification}{options%keyvals}{content}
-\NewTotalTCBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
-\RenewTotalTCBox{\name}{specification}{options%keyvals}{content}
-\RenewTotalTCBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
-\ProvideTotalTCBox{\name}{specification}{options%keyvals}{content}
-\ProvideTotalTCBox[init options%keyvals]{\name}{specification}{options%keyvals}{content}#*
+\ProvideTColorBox[init options]{envname}{specification}{options%keyvals}#*N
```

Comment:
Personally i find adding all variants to the completer exaggerated. It clutters the completion list without any real benefit. None of the commands are used regularly. Completing the general one and adapting it to your needs seems more appropriate (see the reason why now there is a plaintex.cwl ...)

Reply:
Furthermore there has not been any complaint in years that variants are missing ...

Reply:
> Personally i find adding all variants to the completer exaggerated.

Yes. I suggest to only add all variants for `\newtcolorbox` and `\newtcbox`, and removing others. Is this acceptable?

> [...] (see the reason why now there is a plaintex.cwl ...)

That's different. Formats are generally exclusive with each other, and are more basic than packages. It's common that a tex document uses several packages, but one cannot use both formats without a lot of tricks. That's the reason I suggested to move plaintex only macros from auto-loaded `tex.cwl` to manually loaded `plaintex.cwl`. Under ideal conditions, switching between formats should contain both adding and removing operations on completion info, but loading packages almost contains adding operation only.

Given another example, if someday users of ConTeXt format want TeXstudio to add support, there could be a new `context.cwl` which can be manually loaded, instead of adding ConTeXt-only macros to `tex.cwl`.

Reply:
My point is rather clutter which makes it more difficult to select the wanted command, not so much the concrete moving of commands to plaintex.cwl

I would suggest that you really look at it from a practical point of view, so the less variants the better. 
I will accept your suggestion, but maybe you try it out in practice to see how it works ?

Reply:
My motivation is to have texstudio recognizing legal use of commands. Although all variants are added, many of them are suffixed with `#*` hence will show in "all" tag only. And more variants can be marked with `#*`, to leave only the no-optional-arg and full-opt-arg variants in the list. Does this still seems to clutter the completion list, from your point of view?

Reply:
Adding all variants to the cwl does not help the syntax checker. Only the longest variant is internally taken for syntax checking. In case of tcolorbox syntax checking does not work reliably because the commands violate the basic syntax grammar (optional args then mandatory ones) In order to handle the package's grammar, internal storage of commands would need to be changed ...

So, the commands with all arguments need to be there, possibly with #*, and maybe a more usual variant. Putting in all variants is not helpful... (clutter, even with #*)

Reply:
Txs understands optional arguments and how they can be used, though not after mandatory args ...

Reply:
> Txs understands optional arguments and how they can be used, though not after mandatory args ...

Understood.